Zend/zend_enum.c

```diff
@@ -0,0 +1,338 @@
+	return &EG(uninitialized_zval);
+}
+
+static ZEND_COLD void zend_enum_unset_property(zend_object *object, zend_string *member, void **cache_slot)
+{
+	ZEND_ENUM_PROPERTY_ERROR();
+}
+
+ZEND_API zval *zend_enum_get_property_ptr_ptr(zend_object *zobj, zend_string *name, int type, void **cache_slot)
+{
+	return NULL;
+}
+
+int zend_enum_compare_objects(zval *o1, zval *o2)
+{
+	if (Z_TYPE_P(o1) != IS_OBJECT || Z_TYPE_P(o2) != IS_OBJECT) {
+		return ZEND_UNCOMPARABLE;
+	}
+
+	return Z_OBJ_P(o1) == Z_OBJ_P(o2) ? 0 : 1;
+}
+
+void zend_register_enum_ce(void)
+{
+	memcpy(&enum_handlers, &std_object_handlers, sizeof(zend_object_handlers));
+	enum_handlers.read_property = zend_enum_read_property;
+	enum_handlers.write_property = zend_enum_write_property;
+	enum_handlers.unset_property = zend_enum_unset_property;
+	enum_handlers.get_property_ptr_ptr = zend_enum_get_property_ptr_ptr;
+	enum_handlers.clone_obj = NULL;
+	enum_handlers.compare = zend_enum_compare_objects;
+}
+
+void zend_enum_add_interfaces(zend_class_entry *ce)
+{
+	uint32_t num_interfaces_before = ce->num_interfaces;
+
+	ce->num_interfaces++;
+	if (ce->enum_scalar_type != IS_UNDEF) {
+		ce->num_interfaces++;
+	}
+
+	ce->interface_names = erealloc(ce->interface_names, sizeof(zend_class_name) * ce->num_interfaces);
+
+	ce->interface_names[num_interfaces_before].name = zend_string_init("UnitEnum", sizeof("UnitEnum") - 1, 0);
+	ce->interface_names[num_interfaces_before].lc_name = zend_string_init("unitenum", sizeof("unitenum") - 1, 0);
+
+	if (ce->enum_scalar_type != IS_UNDEF) {
+		ce->interface_names[num_interfaces_before + 1].name = zend_string_init("ScalarEnum", sizeof("ScalarEnum") - 1, 0);
+		ce->interface_names[num_interfaces_before + 1].lc_name = zend_string_init("scalarenum", sizeof("scalarenum") - 1, 0);	
+	}
+}
+
+static ZEND_NAMED_FUNCTION(zend_enum_cases_func)
+{
+	zend_class_entry *ce = execute_data->func->common.scope;
+	zend_class_constant *c;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		RETURN_THROWS();
+	}
+
+	array_init(return_value);
+
+	ZEND_HASH_FOREACH_PTR(&ce->constants_table, c) {
+		if (!(c->const_flags & ZEND_CLASS_CONST_IS_CASE)) {
+			continue;
+		}
+		zval *zv = &c->value;
+		if (Z_TYPE_P(zv) == IS_CONSTANT_AST) {
+			zval_update_constant_ex(zv, c->ce);
+			if (UNEXPECTED(EG(exception) != NULL)) {
+				return;
+			}
+		}
+		if (ce->enum_scalar_type == IS_UNDEF) {
+			zend_hash_next_index_insert(Z_ARRVAL_P(return_value), zv);
+		} else {
+			zend_object *zobj = Z_OBJ_P(zv);
+			zend_string *value_string = zend_string_init("value", strlen("value"), 0);
+			zval tmp;
+			zval *value_zv = zend_read_property_ex(ce, zobj, value_string, 0, &tmp);
+			zend_string_free(value_string);
+
+			if (ce->enum_scalar_type == IS_LONG) {
+				zend_hash_index_update(Z_ARRVAL_P(return_value), Z_LVAL_P(value_zv), zv);
+			} else {
+				ZEND_ASSERT(ce->enum_scalar_type == IS_STRING);
+				zend_symtable_update(Z_ARRVAL_P(return_value), Z_STR_P(value_zv), zv);
```

Comment:
Can zend_hash_update here. Though the question of whether the keys are supposed to be there or not remains.